File: src/utils/reminders.js
// Pure decision for the local daily reminder. There is no backend/push service
// (hard constraint), so reminders are best-effort: while the app is open we
// nudge once a day if you're past your reminder hour and haven't studied yet.

export const STORAGE_KEY = 'italian-bible-reminders';
export const DEFAULT_HOUR = 19; // 7pm

// Should we fire the reminder right now? Pure so it can be unit-tested.
export function shouldNotify(prefs, { now = new Date(), studiedToday = false, today } = {}) {
  if (!prefs || !prefs.enabled) return false;
  if (prefs.permission !== 'granted') return false;
  if (studiedToday) return false;
  if (prefs.lastNotified && prefs.lastNotified === today) return false;
  const hour = typeof prefs.hour === 'number' ? prefs.hour : DEFAULT_HOUR;
  return now.getHours() >= hour;
}

File: ios-native/BibbiaCore/Sources/BibbiaCore/Reminders.swift
import Foundation

// Port of src/utils/reminders.js — the pure decision for the local daily
// study reminder. On iOS the app also schedules a real UNUserNotificationCenter
// notification (something the web PWA cannot do), but the in-app nudge logic
// stays identical and unit-tested.

public let DEFAULT_REMINDER_HOUR = 19  // 7pm

public struct ReminderPrefs: Codable, Equatable {
    public var enabled: Bool
    public var hour: Int?
    public var permissionGranted: Bool
    public var lastNotified: String?   // "YYYY-MM-DD"

    public init(enabled: Bool = false, hour: Int? = nil,
                permissionGranted: Bool = false, lastNotified: String? = nil) {
        self.enabled = enabled
        self.hour = hour
        self.permissionGranted = permissionGranted
        self.lastNotified = lastNotified
    }
}

/// Should we fire the reminder right now? Mirrors shouldNotify().
public func shouldNotify(prefs: ReminderPrefs?,
                         now: Date = Date(),
                         studiedToday: Bool,
                         today: String,
                         calendar: Calendar = .current) -> Bool {
    guard let prefs, prefs.enabled else { return false }
    guard prefs.permissionGranted else { return false }
    if studiedToday { return false }
    if let lastNotified = prefs.lastNotified, lastNotified == today { return false }
    let hour = prefs.hour ?? DEFAULT_REMINDER_HOUR
    return calendar.component(.hour, from: now) >= hour
}
